### PR TITLE
Lock free sliding window for internal CircutBreaker metrics

### LINKED ISSUE #2151: CircuitBreakerMetrics performance issue
While doing load testing, I found that metrics.record had blocking issues. Besides marking it as synchronized, can we make a lock-free implementation for it?
 
image

image

### PR DESCRIPTION
**Pull request by flore77**
Fixes #2151.

This PR introduces lock-free implementations of the count-based and time-based sliding windows used internally by the CircuitBreaker. The changes are organized into five commits for easier review:

- **Measurement small refactor**: A small refactor to provide a common interface and accommodate a new measurement implementation.
- **Packed Measurement Implementation**: Added a compact measurement class designed for use with the lock-free sliding windows.
- **Lock-Free Count Sliding Window**: Implemented the lock-free version of the count-based sliding window.
- **Lock-Free Time Sliding Window**: Implemented the lock-free version of the time-based sliding window.
- **New Public Configuration**: Introduced a new configuration option, slidingWindowSynchronizationStrategy, allowing users to select between SYNCHRONIZED (default) and LOCK_FREE.

The new sliding window implementations are validated using existing unit tests, which were extended to include the new versions. Additionally, the correctness of the algorithms - specifically, their linearizability - is verified using the [Lincheck](https://github.com/JetBrains/lincheck) framework.

One note is that the new time sliding window uses a different Clock implementation in order to have access to a monotonic time source, so that it will not be affected by system clock changes. This implementation may yield a better accuracy, which would make it valauble to extend its usage to the whole circuit breaker logic. (most probably as an addendum, to not extend this PR's scope).

This contribution has been made possible also with support from @raresvla.

### CODE REVIEW

**Comment by RobWin** [comment]
Thank you for the PR

**Comment by flore77** [comment]
I also added a JMH benchmark with the following results, on a machine with 16 CPU cores:
```
MetricsBenchmark.benchmarkLFSTW1Thread   thrpt    3   372.161 ±  23.687  ops/ms
MetricsBenchmark.benchmarkLFSTW4Thread   thrpt    3  1404.384 ±  13.286  ops/ms
MetricsBenchmark.benchmarkLFSTW8Thread   thrpt    3  2586.145 ±  38.467  ops/ms
MetricsBenchmark.benchmarkLFSTW16Thread  thrpt    3  3233.989 ± 138.650  ops/ms

MetricsBenchmark.benchmarkSTW1Thread     thrpt    3   372.186 ±  13.524  ops/ms
MetricsBenchmark.benchmarkSTW4Thread     thrpt    3  1409.095 ±  21.550  ops/ms
MetricsBenchmark.benchmarkSTW8Thread     thrpt    3  1646.329 ± 441.343  ops/ms
MetricsBenchmark.benchmarkSTW16Thread    thrpt    3  1513.308 ± 279.168  ops/ms

MetricsBenchmark.benchmarkLFSW1Thread    thrpt    3   373.815 ±   7.472  ops/ms
MetricsBenchmark.benchmarkLFSW4Threads   thrpt    3  1383.414 ±  19.801  ops/ms
MetricsBenchmark.benchmarkLFSW8Threads   thrpt    3  2455.388 ±  45.755  ops/ms
MetricsBenchmark.benchmarkLFSW16Threads  thrpt    3  2496.513 ± 366.928  ops/ms

MetricsBenchmark.benchmarkSW1Thread      thrpt    3   373.069 ±   4.543  ops/ms
MetricsBenchmark.benchmarkSW4Threads     thrpt    3  1420.347 ± 108.867  ops/ms
MetricsBenchmark.benchmarkSW8Threads     thrpt    3  2400.009 ± 215.065  ops/ms
MetricsBenchmark.benchmarkSW16Threads    thrpt    3  1747.785 ± 152.901  ops/ms
```

The results show that the lock-free version performs better when the number of threads increases.

File: resilience4j-core/build.gradle
[the remark this answers is not in the document]
Comment (flore77) [reply resilience4j-core/build.gradle:1]: Hi! I included the Kotlin plugin for the Lincheck tests, as the whole framework and their documentation is Kotlin oriented and felt more natural.

**Review by RobWin — APPROVED** [review]

**Comment by flore77** [comment]
@RobWin thank you for reviewing and merging the PR! 🙇 

When releasing a new version, it could be helpful to mention in the release notes that users opting for the lock-free version should monitor its behavior closely. Concurrency issues are often elusive and tend to manifest under heavy load, so encouraging users to report any anomalies could help us identify and address potential subtle bugs (if any of course 😄).

Also when should we expect a new version to be released? Thanks!

**Pull request merged**